dali/util/custream.cc

```diff
@@ -0,0 +1,51 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/util/custream.h"
+#include "dali/util/cuda_utils.h"
+#include "dali/util/device_guard.h"
+
+namespace dali {
+
+CUStream::CUStream(int device_id, bool default_stream) : stream_{0} {
+  if (!default_stream) {
+    DeviceGuard dg(device_id);
+    CUDA_CALL(cudaStreamCreateWithFlags(&stream_, cudaStreamNonBlocking));
+  }
+}
+
+
+CUStream::~CUStream() {
+  CUDA_CALL(cudaStreamDestroy(stream_));
+}
+
+
+CUStream::CUStream(CUStream &&other) :
+        stream_{other.stream_} {
+  other.stream_ = 0;
+}
+
+
+CUStream &CUStream::operator=(CUStream &&other) {
+  stream_ = other.stream_;
```

Comment:
You're leaking the original `stream_` here. It should be properly destroyed. One way to achieve it is:
```
std::swap(stream_, other.stream_);
other.~CUStream();
```
The above works even in case of self-assignment.

Reply:
Done